src/lib.rs

```diff
@@ -572,24 +566,44 @@ fn derive_where_internal(attr: TokenStream, item: TokenStream) -> Result<TokenSt
     } = syn::parse2(item)?;
 
     // Every trait needs a separate implementation.
-    for trait_ in derive_where.traits() {
+    for trait_ in derive_where.traits {
         let body = trait_.generate_body(&ident, &data)?;
-        let trait_ = trait_.type_();
+        let trait_ = trait_.path();
 
         // Build necessary generics to construct the implementation item.
-        let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
-        // TODO: don't convert to `TokenStream`, but actually properly insert bounds.
-        let mut where_clause = where_clause.map(|where_clause| where_clause.into_token_stream());
+        let (impl_generics, type_generics, mut where_clause) = generics.split_for_impl();
+
+        let default_where = &WhereClause {
+            where_token: Where::default(),
+            predicates: Punctuated::default(),
+        };
+
+        let mut where_clause = where_clause.get_or_insert(default_where).clone();
 
         // If there are any bounds, insert them into the `where` clause.
-        if let DeriveWhere::WithGenerics(generics, _) = &derive_where {
-            // If there is no `where` clause, make one.
-            let where_clause = where_clause.get_or_insert(quote! { where });
-            let bounds = generics.iter().map(|generic| match generic {
-                Generic::CoustomBound(type_bound) => quote! {#type_bound},
-                Generic::NoBound(type_) => quote! {#type_: #trait_},
-            });
-            *where_clause = quote! { #where_clause #(#bounds),* };
+        for generic in &derive_where.generics {
+            where_clause
+                .predicates
+                .push(WherePredicate::Type(match generic {
+                    Generic::CoustomBound(type_bound) => type_bound.clone(),
+                    Generic::NoBound(type_) => {
+                        let mut bounds = Punctuated::default();
+
+                        bounds.push(TypeParamBound::Trait(TraitBound {
+                            paren_token: None,
+                            modifier: syn::TraitBoundModifier::None,
+                            lifetimes: None,
+                            path: trait_.clone(),
+                        }));
```

Comment:
```suggestion
let bounds = Punctuated::from_iterator(Some(
  TypeParamBound::Trait(TraitBound {
    paren_token: None,
    modifier: TraitBoundModifier::None,
    lifetimes: None,
    path: trait_.clone(),
  })
));
```
To avoid unnecessary allocations.
Also import `TraitBoundModifier`.